pkg/monitor/lease.go

```diff
@@ -0,0 +1,179 @@
+	return false, err
+}
+
+func leaseVIPs(cfgPath string, clusterName string, vips []VIP) error {
+	for _, vip := range vips {
+		vipFullName := getInterfaceName(clusterName, vip.name)
+
+		vipMasterIface, _, err := config.GetInterfaceAndNonVIPAddr([]net.IP{vip.ip})
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"name": vip.name,
+				"ip":   vip.ip,
+			}).WithError(err).Error("Failed to get the master device for a vip")
+			return err
+		}
+
+		if err := leaseVIP(cfgPath, vipMasterIface.Name, vipFullName); err != nil {
+			log.WithFields(logrus.Fields{
+				"name":         vipFullName,
+				"masterDevice": vipMasterIface.Name,
+				"vip":          vip,
+			}).WithError(err).Error("Failed to lease a vip")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func leaseVIP(cfgPath, masterDevice, name string) error {
+	iface, err := leaseInterface(masterDevice, name)
+
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"masterDevice": masterDevice,
+			"name":         name,
+		}).WithError(err).Error("Failed to lease interface")
+		return err
+	}
+
+	cmd := exec.Command("dhclient", "-d", "--no-pid", "-sf", "/bin/true",
+		"-lf", getLeaseFile(cfgPath, name), "-v", iface.Name, "-H", name)
+
+	return cmd.Start()
+}
+
+func leaseInterface(masterDevice string, name string) (*net.Interface, error) {
+	mac := generateMac(macPrefixQumranet[:], name)
+
+	// Check if already exist
+	if macVlanIfc, err := net.InterfaceByName(name); err == nil {
+		return macVlanIfc, nil
+	}
+
+	// Read master device
+	master, err := netlink.LinkByName(masterDevice)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"masterDev": masterDevice,
+		}).WithError(err).Error("Failed to read master device")
+		return nil, err
+	}
+
+	linkAttrs := netlink.LinkAttrs{
+		Name:         name,
+		ParentIndex:  master.Attrs().Index,
+		HardwareAddr: mac,
+	}
+
+	mv := &netlink.Macvlan{
+		LinkAttrs: linkAttrs,
+		Mode:      netlink.MACVLAN_MODE_BRIDGE,
+	}
+
+	// Create interface
+	if err := netlink.LinkAdd(mv); err != nil {
+		log.WithFields(logrus.Fields{
+			"masterDev": masterDevice,
+			"name":      name,
+		}).WithError(err).Error("Failed to create a macvlan")
+		return nil, err
+	}
+
+	// Read created link
+	macvlanInterfaceLink, err := netlink.LinkByName(name)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"name": name,
+		}).WithError(err).Error("Failed to read new device")
+		return nil, err
+	}
+
+	// Bring the interface up
+	if err = netlink.LinkSetUp(macvlanInterfaceLink); err != nil {
+		log.WithFields(logrus.Fields{
+			"interface": name,
+		}).WithError(err).Error("Failed to bring interface up")
+		return nil, err
+	}
+
+	// Read created interface
+	macVlanIfc, err := net.InterfaceByName(name)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"name": name,
+		}).WithError(err).Error("Failed to read new device")
+		return nil, err
+	}
+
+	return macVlanIfc, nil
+}
+
+func generateMac(prefix []byte, noise string) (mac net.HardwareAddr) {
```

Comment:
Nit: I wouldn't call this "noise". That suggests a randomness that may or may not be wanted. It's just a key passed to a hash function, so I would call it "key".